cmd/erasure-createfile.go

```diff
@@ -18,85 +18,93 @@ package cmd
 
 import (
 	"context"
-	"hash"
 	"io"
 
+	"sync"
+
 	"github.com/minio/minio/cmd/logger"
 )
 
-// CreateFile creates a new bitrot encoded file spread over all available disks. CreateFile will create
-// the file at the given volume and path. It will read from src until an io.EOF occurs. The given algorithm will
-// be used to protect the erasure encoded file.
-func (s *ErasureStorage) CreateFile(ctx context.Context, src io.Reader, volume, path string, buffer []byte, algorithm BitrotAlgorithm, writeQuorum int) (f ErasureFileInfo, err error) {
-	if !algorithm.Available() {
-		logger.LogIf(ctx, errBitrotHashAlgoInvalid)
-		return f, errBitrotHashAlgoInvalid
+// Writes in parallel to bitrotWriters
+type parallelWriter struct {
+	writers     []*bitrotWriter
+	writeQuorum int
+}
+
+// Append appends data to bitrotWriters in parallel.
+func (p *parallelWriter) Append(ctx context.Context, blocks [][]byte) error {
+	var wg sync.WaitGroup
+	errs := make([]error, len(p.writers))
+	for i, writer := range p.writers {
+		if writer == nil {
+			errs[i] = errDiskNotFound
+			continue
+		}
+
+		wg.Add(1)
+		go func(i int, writer *bitrotWriter) {
+			defer wg.Done()
+			errs[i] = writer.Append(blocks[i])
+			if errs[i] != nil {
+				p.writers[i] = nil
+			}
+		}(i, writer)
 	}
-	f.Checksums = make([][]byte, len(s.disks))
-	hashers := make([]hash.Hash, len(s.disks))
-	for i := range hashers {
-		hashers[i] = algorithm.New()
+	wg.Wait()
+
+	// If nilCount >= p.writeQuorum, we return nil.
+	// HealFile() uses CreateFile with p.writeQuorum=1 to accommodate healing of single disk.
+	nilCount := 0
+	for _, err := range errs {
+		if err == nil {
+			nilCount++
+		}
 	}
-	errChans, errs := make([]chan error, len(s.disks)), make([]error, len(s.disks))
-	for i := range errChans {
-		errChans[i] = make(chan error, 1) // create buffered channel to let finished go-routines die early
+	if nilCount >= p.writeQuorum {
+		return nil
 	}
+	return reduceWriteQuorumErrs(ctx, errs, objectOpIgnoredErrs, p.writeQuorum)
+}
 
-	var blocks [][]byte
-	var n = len(buffer)
-	for n == len(buffer) {
-		n, err = io.ReadFull(src, buffer)
-		if n == 0 && err == io.EOF {
-			if f.Size != 0 { // don't write empty block if we have written to the disks
+// CreateFile reads from the reader, erasure-encodes the data and writes to the writers.
+func (s *ErasureStorage) CreateFile(ctx context.Context, src io.Reader, writers []*bitrotWriter, buf []byte, quorum int) (total int64, err error) {
+	writer := &parallelWriter{
+		writers:     writers,
+		writeQuorum: quorum,
+	}
+	for {
+		eof := false
+		var blocks [][]byte
+		n, err := io.ReadFull(src, buf)
+		if err != nil && err != io.EOF && err != io.ErrUnexpectedEOF {
+			logger.LogIf(ctx, err)
+			return 0, err
+		}
+		if err == io.EOF || err == io.ErrUnexpectedEOF {
```

Comment:
Suggestion: replace by `eof := err == io.EOF || err == io.ErrUnexpectedEOF`
Then `eof := false` not needed...